Bufferline timestamps now try to detect locale

pmelanson opened the pull request:
Follows the same pattern as `injectDateChangeMessageIfNeeded` in file `js/handlers.js`, roughly.

Adds a bit more complexity to figure out what the user's locale time string is.

Also, moves a little bit of code complexity from index.html to js/models.js . The reasoning behind this is the format of the time string will only be figured out when javascript code is being executed, and if we want to apply a different style to the time separators as we do now, then we'll have to do that in the JS code.

Prototype PR, so for sure let me know if there's anything to change. For example, [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleTimeString#Performance) suggests I can use `Intl.DateTimeFormat` something or other to get better performance for large amounts of dates.

Thoughts?

pmelanson commented on css/glowingbear.css:
```diff
@@ -326,6 +326,7 @@ tr.bufferline {
 td.time {
     padding: 1px 5px 1px 1px;
     vertical-align: top;
+    text-align: right;
```
having a 12-hour time locale stops the hour field returned from `Date.toLocaleTimeString()` from being zero-padded, so if we have something like:

```
9:59 pm [text]
10:00 pm [text]
```

This change makes the above turn into:


```
 9:59 pm [text]
10:00 pm [text]
```

pmelanson commented:
Fixes #782

pmelanson commented:
oh man totally forgot to keep support for the seconds option, I'll get that tomorrow X/

lorenzhs commented on index.html:
```diff
@@ -289,7 +289,7 @@ <h4 class="panel-title">
               <td class="time">
                 <span class="date" ng-class="::{'repeated-time': bufferline.shortTime==bufferlines[$index-1].shortTime}">
-                  <span class="cof-chat_time cob-chat_time coa-chat_time" ng-bind="::(bufferline.date|date:'HH')"></span><span class="cof-chat_time_delimiters cob-chat_time_delimiters coa-chat_time_delimiters">:</span><span class="cof-chat_time cob-chat_time coa-chat_time" ng-bind="::(bufferline.date|date:'mm')"></span><span class="seconds"><span class="cof-chat_time_delimiters cob-chat_time_deli ...
+                  <span class="cof-chat_time cob-chat_time coa-chat_time" ng-bind-html="bufferline.formattedTime"></span>
```
Can you bind this once with the "::" syntax? Reduces watchers and improves performance

pmelanson commented:
addressed. lmk if my angular flailings are wrong again.

lorenzhs commented on index.html:
```diff
@@ -288,8 +288,9 @@ <h4 class="panel-title">
-                  <span class="cof-chat_time cob-chat_time coa-chat_time" ng-bind="::(bufferline.date|date:'HH')"></span><span class="cof-chat_time_delimiters cob-chat_time_delimiters coa-chat_time_delimiters">:</span><span class="cof-chat_time cob-chat_time coa-chat_time" ng-bind="::(bufferline.date|date:'mm')"></span><span class="seconds"><span class="cof-chat_time_delimiters cob-chat_time_deli ...
+                <span ng-if="settings.showtimestamp" class="date" ng-class="::{'repeated-time': bufferline.shortTime==bufferlines[$index-1].shortTime}">
+                  <span ng-if="!settings.showtimestampSeconds" class="cof-chat_time cob-chat_time coa-chat_time" ng-bind-html="::bufferline.formattedTime"></span>
+                  <span ng-if="settings.showtimestampSeconds" class="cof-chat_time cob-chat_time coa-chat_time" ng-bind-html="::bufferline.formattedTimeSeconds"></span>
```
These `ng-if`s introduce 2 watchers per line - I suggest just refreshing the buffer when that setting is changed (as a setting change hook) and using the single-binding syntax (`::`) here as well

lorenzhs commented on index.html:
```diff
@@ -276,7 +276,7 @@ <h4 class="panel-title">
           </ul>
         </div>
-        <table ng-class="{'notimestamp':!settings.showtimestamp,'notimestampseconds':!settings.showtimestampSeconds}">
+        <table>
```
I would suggest you keep this around and use classes on the `span`s to show/hide instead of `ng-if`s

pmelanson commented:
Changes made. I haven't run tests on this yet. Reverted to what was happening before with classes, also added in the thing Techman- asked about with zero padding.

lorenzhs commented:
Thanks, that solved the issue with the watchers. However the right-alignment changes the view a little bit compared to how it was previously. Timestamps are a bit further right in 24h format, creating some wasted space to the left of them. Do you think you could change that?

lorenzhs commented:
Hmm it misclassifies me (showing 12h format instead of 24h), my browser language settings are (in order of priority, descending): English (UK), English, German, English (US). Got any idea why?

pmelanson commented:
Removed the right align rule. Had trouble getting chrome to recognize locales, but here's a screenshot of what it looks like now:

![no-align](https://cloud.githubusercontent.com/assets/1498589/16172876/b8c4adc6-355f-11e6-9d3e-8acc541151a5.png)

Techman- commented:
So, what's the status on this?

pmelanson commented:
I think I've fixed everything, and the original person who asked for this feature is satisfied. I've tested it with US locale (which is 12h) and Canada locale (24h) and it looks like it changes. No real-world testing though.

I think it's ready but that's just me

lorenzhs commented:
Well then let's just merge it and see if anyone complains :) Thanks @pmelanson and sorry about the delay.
